### DISCUSSION TITLE
Initialization / Factory Methods and Objects

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5152] [Language Ideas]
For C#11, I noticed that object initialization was one of the things on the roadmap.

I would like the ability to have a function that would serve as a "secondary" constructor and allow me to use it with object initializer syntax. 

TLDR:
```
public class Person {
  public long Id {get; set;}
  public string Name {get; set; }
  public int Age {get; set;} 
  ...
}

[Factory]
public Person FromSession(Session s){
//...
}

//This is called similar to an object initializer.
var P = FromSession(...) {
  Id = 23,
};

```

**Comment by bernd5** [comment]
It should work not just with "set" - it should work with "init", too.
And to simplify the handling I would recommend a "with expression".

**Comment by 333fred** [comment]
This would be the factory methods that we mentioned.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2020/LDM-2020-04-20.md

**Comment by Joe4evr** [comment]
You should watch this: https://youtu.be/fIXujTOA6BE?t=50m

**Comment by TonyValenti** [comment]
@Joe4evr Thanks for sharing this!  This is really close to what I'd like.  Here are some thoughts for LDM team that I'd like to pass along:

1.   Recycling ```with```
I feel like ```with``` already has "Make a copy of this" semantics. Because of this, a different keyword should be used.

2.  Compatibility with existing code
I don't want to need to decorate existing classes and methods with [Factory] to opt-in to this feature.  It should simply just be available.

3.  Including Mutation and not just Initialization
Initializer syntax is really nice and crisp.  Why not allow it for mutation too?

Here are some thoughts on how the code might look:
```
var p = Person.Create() set {
  FirstName = "John",
  LastName = "Smith",
};

p set {
  Age = 21,
  Relatives += { Person.Create() }
  OnSave += EventHandler,
};

p set {
  Age = 91,
  Relatives -= { DeceasedPerson }
  OnSave -= EventHandler,
};

```

You'll notice a few things:
1.  ```set``` used as the 'don't make a copy` version of ```with```
2. It can be used in more places than just construction
3. It can wire up event handlers and add items to a collection using ```+=```
4. It can detach event handlers and remove items from a collection using ```-=```

**Reply by Joe4evr** [reply]
> 1. `set` used as the 'don't make a copy` version of `with`

For your first example, you wouldn't even need a new keyword. An object initializer after a factory method will do:
```cs
var p = Person.Create()
{
    FirstName = "John",
    LastName = "Smith",
};
```
> 2. It can be used in more places than just construction

That's something the language team wants to avoid, tho I don't know the precise reason why.

> 3. It can wire up event handlers and add items to a collection using `+=`

I do believe adding event handlers in an object initializer is already something the language team is considering. Adding items to collections inside an initializer is already supported and doesn't need an operator:
```cs
var p = new Person()
{
    Relatives = { new Person() } // provided the collection isn't null
};
```

> 4. It can detach event handlers and remove items from a collection using `-=`

As mentioned with point 2, it's extremely unlikely C# will use object initializer syntax to modify existing instances, so this is a non-starter since a newly created object is unlikely to have any event handlers attached to it.
